---
title: "Low-quality Cells Validation"
output: html_document
---

```{r global_options, include=FALSE}
library(ggplot2)
library(ggpubr)
library(ggrastr)
library(dplyr)
library(magrittr)
library(Matrix)
library(parallel)
library(reshape2)
library(pagoda2)

source("./Functions/PlotFuncs.R")
source("./Functions/Functions.R")
source("./Functions/PagodaWrappers.R")

knitr::opts_chunk$set(fig.width=5, fig.height=3, echo=FALSE, warning=FALSE, message=FALSE)

theme_set(theme_base)

set.seed(42)
kDataPath <- '/d0-mendel/home/viktor_petukhov/Data/allon_new/'
kDatasetName <- 'SRR387961_'

kPlotsFolder <- paste0('/d0-mendel/home/viktor_petukhov/Data/Plots/PaperReview/LowQualityCells/', kDatasetName, '/')
kCsvLink <- 'ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2230nnn/GSM2230762/suppl/GSM2230762%5Fmouse2%5Fumifm%5Fcounts%2Ecsv%2Egz'
kRescuedScoreThreshold <- 0.9
kFilteredScoreThreshold <- 0.1
# kScoreThreshold <- 0.9
```

```{r}
LibrariesUnion <- function(libs, names.func=names, set.names.func=setNames, union.func=unlist) {
  return(lapply(1:length(libs), function(i) set.names.func(libs[[i]], paste0(names.func(libs[[i]]), '-', i))) %>% union.func())
}
```

## Pipeline data
```{r}
rds_paths <- paste0(kDataPath, c('SRR3879617/est_11_22/', 'SRR3879618/est_11_22/', 'SRR3879619/est_11_29/'), 'cell.counts.rds')
holders <- mclapply(rds_paths, readRDS, mc.cores=3)
```

```{r}
cms <- lapply(holders, `[[`, 'cm')
common_genes <- Reduce(intersect, lapply(cms, rownames))
cms <- lapply(cms, function(x) x[common_genes,])

cm_union <- LibrariesUnion(cms, colnames, `colnames<-`, function(x) Reduce(cbind, x))
scores <- mclapply(holders, ScorePipelineCells, mit.chromosome.name='chrM', mc.cores=length(holders)) %>% LibrariesUnion()

mit_fracs <- lapply(holders, `[[`, 'reads_per_chr_per_cells') %>% lapply(`[[`, 'Exon') %>%
  lapply(GetChromosomeFraction, 'chrM') %>% LibrariesUnion()

umis_per_cb <- sort(Matrix::colSums(cm_union), decreasing=T)
```

## Data from the paper
```{r}
published_csv <- url(kCsvLink) %>% gzcon() %>% readLines() %>% paste0(collapse='\n') %>% data.table::fread(data.table=F)

published_csv <- published_csv[, 1:3] %>%
  mutate(LibId = substr(V1, 11, 11) %>% as.integer(),
         Barcode = gsub("-", "", barcode) %>% paste0('-', LibId),
         Cluster = gsub("_", " ", assigned_cluster)) %>%
  dplyr::select(-V1, -barcode, -LibId, -assigned_cluster)

cluster_by_barcodes <- setNames(published_csv$Cluster, published_csv$Barcode)
dim(published_csv)
```

## Pagoda analysis
```{r}
r_cm <- cm_union[, order(Matrix::colSums(cm_union), decreasing=T)]

cell_mask <- setNames(rep(FALSE, ncol(r_cm)), colnames(r_cm))
cell_mask[names(scores)[scores > kRescuedScoreThreshold]] <- TRUE
cell_mask[intersect(names(cluster_by_barcodes), names(cell_mask))] <- TRUE
r_cm <- r_cm[, cell_mask]

r_cm <- r_cm[Matrix::rowSums(r_cm) > 20, ]

cluster_by_barcodes <- cluster_by_barcodes[names(cluster_by_barcodes) %>% intersect(colnames(r_cm))]

any(Matrix::colSums(r_cm) == 0)
dim(r_cm)
```

```{r, message=FALSE}
r <- GetPagoda(r_cm, n.cores=10)
```

```{r, fig.width=6, fig.height=4}
PlotPagodaEmbeding(r, colors=mit_fracs, show.legend=T, title='Mitochondrial fraction',
                   alpha=0.9, size=0.5, font.size=4.5) +
  scale_color_continuous(name='Fraction') +
  theme_pdf(legend.pos=c(1, 1))
```

```{r}
filtered_cbs <- intersect(names(cluster_by_barcodes), names(scores)[scores <= kFilteredScoreThreshold])
rescued_cbs <- setdiff(names(scores)[scores > kRescuedScoreThreshold], names(cluster_by_barcodes))
intersect_cbs <- intersect(names(r$clusters$PCA$infomap), names(cluster_by_barcodes))
```

```{r}
# cluster_centers <- split(intersect_cbs, cluster_by_barcodes[intersect_cbs]) %>%
#   lapply(function(cbs) r$reductions$PCA[cbs, ] %>% Matrix::colMeans())
# 
# rescued_clusters <- apply(r$reductions$PCA[rescued_cbs, ], 1, function(cell)
#   sapply(cluster_centers, function(clust) norm(cell - clust, type='2')) %>% which.min() %>% names())
notannotated_cells <- setdiff(colnames(r_cm), names(cluster_by_barcodes))
clusters_annotated_resc <- AnnotateClusters(r$graphs$PCA, cluster_by_barcodes, notannotated_cells,
                                            max.iter=100, mc.cores=10)
rescued_clusters <- clusters_annotated_resc[rescued_cbs]
```

```{r}
# clusts_after_filt <- cluster_by_barcodes[intersect(names(cluster_by_barcodes), names(scores)[scores > kScoreThreshold])]
clusts_after_filt <- cluster_by_barcodes[setdiff(names(cluster_by_barcodes), filtered_cbs)]
real_clusts <- c(clusts_after_filt, rescued_clusters)

cell_types_freqs <- table(cluster_by_barcodes)
data.frame(Raw=as.vector(cell_types_freqs),
           Filtered=as.vector(table(factor(clusts_after_filt, levels=names(cell_types_freqs)))),
           WithRescued=as.vector(table(factor(real_clusts, levels=names(cell_types_freqs)))),
           row.names=names(cell_types_freqs))
```

```{r, message=FALSE, fig.width=4, fig.height=4.6}
plot_clusters <- as.factor(cluster_by_barcodes[names(scores)[scores > kFilteredScoreThreshold]])
gg_tsne <- PlotFiltrationResults(r, plot_clusters, filtered_cbs, factor(rescued_clusters, levels=levels(plot_clusters)),
                                 unchanged.alpha=1, rescued.alpha=1, raster.width=7 / 1.75, raster.height=6 / 1.3) +
  theme_pdf(legend.pos=c(1, 1), show.ticks=F)

gg_tsne
# ggsave(paste0(kPlotsFolder, 'rescued_cells_tsne.pdf'), width=6, height=4)
```

```{r, message=FALSE}
library(Seurat)
srt <- CreateSeuratObject(raw.data = r_cm, min.cells = 10, min.genes = 0)
srt <- NormalizeData(object = srt, normalization.method = "LogNormalize", scale.factor = 10000)
srt <- FindVariableGenes(object = srt, mean.function = ExpMean, dispersion.function = LogVMR,
                         x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 1, do.plot=F)
srt <- ScaleData(object = srt, vars.to.regress = "nUMI")
```

```{r}
srt@ident <- as.factor(clusters_annotated_resc[colnames(srt@raw.data)])
names(srt@ident) <- colnames(srt@raw.data)
# compared_clusters <- unique(srt@ident)
compared_clusters <- c('beta', 'delta', 'alpha', 'gamma')
cluster_markers <- mclapply(compared_clusters, function(i) mclapply(setdiff(compared_clusters, i), FindClusterMarkers, i, srt, mc.cores=3), mc.cores=4)
overexpressed_genes <- GetOverexpressedGenes(srt, compared_clusters, cluster_markers, genes.from.cluster = 50, expression.threshold = 0.3)
length(overexpressed_genes)
```

```{r}
selected_clusters <- compared_clusters

tested_clusts <- sort(real_clusts[real_clusts %in% selected_clusters])
separation <- c(setNames(rep('rescued', length(rescued_cbs)), rescued_cbs),
                setNames(rep('real', length(clusts_after_filt)), names(clusts_after_filt)))

umis_per_cb_subset <- log10(Matrix::colSums(r_cm[, names(tested_clusts)]))
tested_clusts <- tested_clusts[order(tested_clusts, -umis_per_cb_subset)]
de_genes <- intersect(overexpressed_genes, colnames(r$counts))
```

```{r, fig.width=6, fig.height=7}
plot_df <- ExpressionMatrixToDataFrame(r$counts[names(tested_clusts), de_genes], umis_per_cb_subset, tested_clusts, rescued_cbs)
# plot_df <- plot_df %>% filter(UmisPerCb < 3.5)
plot_dfs <- split(plot_df, plot_df$IsRescued) %>% setNames(c('real', 'rescued'))

ggs <- lapply(plot_dfs, HeatmapAnnotGG) %>%
  lapply(lapply, `+`, theme(plot.margin=margin()))

gg_legends <- mapply(`+`, ggs$real, list(HeatmapLegendGuide('Expression'),
                                    HeatmapLegendGuide('Cell type', guide=guide_legend, nrow=4),
                                    HeatmapLegendGuide('log10(#molecules)')), SIMPLIFY=F) %>%
  lapply(`+`, theme(legend.margin=margin(l=4, r=4, unit='pt'))) %>% lapply(get_legend)

ggs$real$heatmap <- ggs$real$heatmap + rremove('xlab') + ylab('Cells')
ggs$rescued$heatmap <- ggs$rescued$heatmap + labs(x = 'Genes', y = 'Cells')
ggs_annot <- lapply(ggs, function(gg) cowplot::plot_grid(plotlist=lapply(gg, `+`, rremove('legend')), nrow=1, rel_widths=c(1.5, 0.1, 0.1), align='h'))

gg_legends_plot <- cowplot::plot_grid(plotlist=gg_legends, nrow=3, align='v')

gg_left <- cowplot::plot_grid(ggs_annot$real, ggs_annot$rescued, nrow=2, labels='AUTO')
```

```{r, fig.width=7, fig.height=6}
gg_legends_plot <- cowplot::plot_grid(plotlist=gg_legends[c(1, 3)], nrow=2, align='v') %>%
  cowplot::plot_grid(gg_legends[[2]], ncol=2, rel_widths=c(2, 1))

gg_right <- cowplot::plot_grid(gg_tsne + theme(plot.margin=margin(l=0.1, unit='in')), gg_legends_plot,
                               nrow=2, rel_heights=c(1, 0.3), align='v', labels=c('C'))

# dev.off()
cowplot::plot_grid(gg_left, gg_right, rel_widths=c(0.75, 1))
ggsave(paste0(kPlotsFolder, 'SRR3_figure.pdf'), width=7, height=6)
```


# Old figure
```{r}
kPlotWidth <- 4
kPlotHeight <- 2.5
kPlotDpi <- 200
kPointSize <- 0.3
```

```{r, fig.width=4, fig.height=3}
gp_clusters <- PlotPagodaEmbeding(r, clustering.type='infomap', show.legend=F, min.cluster.size=10,
                                  mark.clusters=T, alpha=0.5, size=kPointSize, font.size=4.5, raster=T,
                                  raster.width=kPlotWidth, raster.height=kPlotHeight, raster.dpi=kPlotDpi) +
  theme_pdf(F)
gp_clusters
```

```{r, fig.width=4, fig.height=3}
gp_umis_per_cb <- PlotPagodaEmbeding(r, colors=Matrix::colSums(r_cm), show.legend=T, alpha=0.5, size=kPointSize, raster=T,
                                     raster.width=kPlotWidth, raster.height=kPlotHeight, raster.dpi=kPlotDpi) +
  theme_pdf(F, legend.pos=c(0, 0)) +
  guides(color=guide_colorbar(title='#Molecules per cell', direction='horizontal', barwidth=unit(1.5, 'in'), title.position='top', title.hjust=0, raster=T)) +
  scale_color_distiller(trans='log', breaks=c(100, 1000, 5000), palette='YlGn') +
  theme(legend.margin = ggplot2::margin(t=0, r=0, b=0, l=0.05, unit='in'))
gp_umis_per_cb
```

```{r, fig.width=4, fig.height=3}
gp_mit <- PlotPagodaEmbeding(r, colors=mit_fracs, show.legend=T, alpha=0.5, size=kPointSize, show.ticks=F, raster=T,
                             raster.width=kPlotWidth, raster.height=kPlotHeight, raster.dpi=kPlotDpi) +
  scale_color_distiller(values=c(0, 0.05, 0.1, 0.3, 1.0), breaks=seq(0, 0.9, 0.3), palette='RdYlBu') +
  theme_pdf(show.ticks=F, legend.pos=c(0, 0)) +
  guides(color=guide_colorbar(title='Mitochondrial fraction', direction='horizontal', barwidth=unit(1.5, 'in'), title.position='top', raster=T)) +
  theme(legend.margin = ggplot2::margin(t=0, r=0, b=0, l=0.05, unit='in'))
gp_mit
```

```{r, fig.width=4, fig.height=3}
is_real <- as.integer(scores > 0.6)
names(is_real) <- names(scores)

gp_score <- PlotPagodaEmbeding(r, colors=scores, show.legend=T, min.cluster.size=100, alpha=0.5, size=kPointSize, raster=T,
                               raster.width=kPlotWidth, raster.height=kPlotHeight, raster.dpi=kPlotDpi) +
  scale_color_distiller(values=c(1.0, 0.8, 0.7, 0.6, 0.0), palette='RdYlBu', guide=guide_colourbar(title='Score', raster=T), breaks=seq(0.1, 0.9, length.out=3)) +
  theme_pdf(F, c(0, 0)) +
  guides(color=guide_colorbar(title='Score', direction='horizontal', barwidth=unit(1, 'in'), title.position='top', title.hjust=0.5)) +
  theme(legend.margin = ggplot2::margin(t=0, r=0.05, b=0, l=0.05, unit='in'))
gp_score
```

```{r, fig.width=8, fig.height=5}
gg_figure <- cowplot::plot_grid(plotlist=lapply(list(gp_clusters, gp_umis_per_cb, gp_mit, gp_score),
                                                function(x) x + rremove("xylab") + theme(plot.margin=margin(l=0.01, r=0.01, b=0.01, t=0.01, "in"))),
                                labels = c("A", "B", "C", "D"), nrow=2, ncol=2, label_x=0.01, label_y=0.99)

gg_figure <- annotate_figure(gg_figure, bottom=text_grob('tSNE-1', size=14), left=text_grob('tSNE-2', size=14, rot=90))

gg_figure
# ggsave(filename=paste0(kPlotsFolder, 'fig_clusters.pdf'), width=8, height=5, units='in')
```
