<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Viktor Petukhov" />

<meta name="date" content="2018-01-23" />

<title>Drop-seq human/mouse mixture analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">dropEstAnalysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/workflowr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Drop-seq human/mouse mixture analysis</h1>
<h4 class="author"><em>Viktor Petukhov</em></h4>
<h4 class="date"><em>2018-01-23</em></h4>

</div>


<p><strong>Source file:</strong> notebooks/human_mouse/hm_drop_seq.Rmd</p>
<p><strong>Last updated:</strong> 2018-01-27</p>
<p><strong>Code version:</strong> 1485196</p>
<pre class="r"><code>library(ggplot2)
library(ggrastr)
library(dplyr)
library(dropestr)
library(dropEstAnalysis)

theme_set(theme_base)</code></pre>
<div id="load-data" class="section level2">
<h2>Load data</h2>
<p>Here bam file was filtered by realigning it with <code>kallisto 0.43</code> separately on mouse and human genome. Only reads, which were aligned only on one of them were used in <code>dropEst</code>.</p>
<pre class="r"><code>holder &lt;- readRDS(&#39;../../data/dropest/dropseq/thousand/est_2018_01_26_tophat/thousand.rds&#39;)
bc_data &lt;- PrepareLqCellsDataPipeline(holder, scale=FALSE) %&gt;% 
  tibble::rownames_to_column(&quot;CB&quot;)</code></pre>
<pre class="r"><code>cm_real &lt;- holder$cm_raw
cell_number &lt;- 1100

gene_species &lt;- ifelse(substr(rownames(cm_real), 1, 2) == &quot;HU&quot;, &#39;Human&#39;, &#39;Mouse&#39;) %&gt;% 
  as.factor()
umi_by_species &lt;- lapply(levels(gene_species), function(l) cm_real[gene_species == l,] %&gt;% 
                           Matrix::colSums()) %&gt;% as.data.frame() %&gt;% 
  `colnames&lt;-`(levels(gene_species)) %&gt;% tibble::rownames_to_column(&#39;CB&#39;) %&gt;% 
  as_tibble() %&gt;% 
  mutate(Total = Human + Mouse, Organism=ifelse(Human &gt; Mouse, &quot;Human&quot;, &quot;Mouse&quot;), 
         IsReal=rank(Total) &gt;= length(Total) - cell_number) %&gt;%
  filter(Total &gt; 20)

mit_genes &lt;- rownames(holder$cm_raw)[grep(&quot;_MT:.+&quot;, rownames(holder$cm_raw))]
umi_by_species &lt;- umi_by_species %&gt;% mutate(
    MitochondrionFraction = GetGenesetFraction(holder$cm_raw, mit_genes)[CB]
  )</code></pre>
</div>
<div id="common-view" class="section level2">
<h2>Common view</h2>
<pre class="r"><code>gg &lt;- ggplot(umi_by_species, aes(x=Mouse, y=Human)) + 
  geom_abline(aes(slope=1, intercept=0), linetype=&#39;dashed&#39;, alpha=0.5) +
  scale_x_log10(limits=c(1, 2e5), name=&quot;#Mouse molecules&quot;) + 
  scale_y_log10(name=&quot;#Human molecules&quot;) + annotation_logticks() +
  theme_pdf(legend.pos=c(0.97, 0.05)) + theme(legend.margin=margin(l=3, r=3, unit=&quot;pt&quot;))

gg_left &lt;- gg + geom_point(aes(color=IsReal), size=0.1, alpha=0.15) +
  guides(color=guide_legend(override.aes=list(size=1.5, alpha=1)))

gg_right &lt;- gg + geom_point(aes(color=MitochondrionFraction), size=0.1, alpha=0.15) +
  scale_color_gradientn(colours=c(&quot;#1200ba&quot;, &quot;#347fff&quot;, &quot;#cc4000&quot;, &quot;#ff3333&quot;), 
                        values=scales::rescale(c(0, 0.1, 0.3, 0.8))) +
  guides(color=guide_colorbar(direction=&quot;horizontal&quot;, title.position=&quot;top&quot;, 
                              title=&quot;Mitochondrial\nfraction&quot;, 
                              barwidth=unit(1.0, units=&quot;in&quot;)))

cowplot::plot_grid(gg_left, gg_right)</code></pre>
<p><img src="figure/hm_drop_seq.Rmd/unnamed-chunk-4-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggplot(umi_by_species) + 
  geom_point(aes(x=Total, y=pmin(Human, Mouse) / Total, color=Organism), size=0.1, 
             alpha=0.1) +
  scale_x_log10(name=&#39;#Real UMIs&#39;, limits=c(10, 2e5)) + annotation_logticks() + 
  ylab(&#39;Fraction of mixed UMIs&#39;) +
  guides(color=guide_legend(override.aes=list(size=1.5, alpha=1))) +
  theme_pdf(legend.pos=c(1, 1))</code></pre>
<p><img src="figure/hm_drop_seq.Rmd/unnamed-chunk-5-1.png" width="384" style="display: block; margin: auto;" /></p>
</div>
<div id="check-for-constant-background" class="section level2">
<h2>Check for constant background</h2>
<p>Background cells have constant fraction of mouse and human reads. Though this fraction is not determined by the corresponding ratio in real cells:</p>
<pre class="r"><code>mouse_frac &lt;- umi_by_species %&gt;% filter(IsReal) %&gt;% 
  summarise(Mouse=sum(Mouse[Organism == &#39;Mouse&#39;]), Human=sum(Human[Organism == &#39;Human&#39;]), 
            MF=Mouse / (Mouse + Human)) %&gt;% .$MF

ggplot(umi_by_species) + 
  geom_histogram(aes(x=Mouse / Total, y=..density.., fill=IsReal), binwidth=0.005, position=&quot;identity&quot;) + 
  geom_vline(xintercept=mouse_frac) +
  xlab(&quot;Fraction of mouse reads&quot;) +
  theme_pdf(legend.pos=c(1, 1))</code></pre>
<p><img src="figure/hm_drop_seq.Rmd/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Distribution of total number of molecules by background cells:</p>
<pre class="r"><code>gg &lt;- ggplot(umi_by_species %&gt;% filter(!IsReal)) +
  geom_histogram(aes(x=Total), bins=100) +
  scale_x_continuous(limits=c(0, 1000), expand=c(0, 0), name=&quot;Total #UMIs&quot;) +
  scale_y_continuous(limits=c(0,4500), expand=c(0, 0), name=&quot;#Cells&quot;) +
  theme_pdf()

gg</code></pre>
<p><img src="figure/hm_drop_seq.Rmd/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Though, it looks better on logscale:</p>
<pre class="r"><code>gg + 
  scale_x_log10(limits=c(19, 3000), expand=c(0, 0), name=&quot;Total #UMIs&quot;) +
  annotation_logticks(sides=&quot;b&quot;)</code></pre>
<p><img src="figure/hm_drop_seq.Rmd/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>data.frame(value=unlist(sessioninfo::platform_info()))</code></pre>
<div class="kable-table">
<table>
<thead>
<tr class="header">
<th></th>
<th align="left">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>version</td>
<td align="left">R version 3.4.1 (2017-06-30)</td>
</tr>
<tr class="even">
<td>os</td>
<td align="left">Ubuntu 14.04.5 LTS</td>
</tr>
<tr class="odd">
<td>system</td>
<td align="left">x86_64, linux-gnu</td>
</tr>
<tr class="even">
<td>ui</td>
<td align="left">X11</td>
</tr>
<tr class="odd">
<td>language</td>
<td align="left">(EN)</td>
</tr>
<tr class="even">
<td>collate</td>
<td align="left">en_US.UTF-8</td>
</tr>
<tr class="odd">
<td>tz</td>
<td align="left">America/New_York</td>
</tr>
<tr class="even">
<td>date</td>
<td align="left">2018-01-27</td>
</tr>
</tbody>
</table>
</div>
<pre class="r"><code>as.data.frame(sessioninfo::package_info())[c(&#39;package&#39;, &#39;loadedversion&#39;, &#39;date&#39;, &#39;source&#39;)]</code></pre>
<div class="kable-table">
<table>
<thead>
<tr class="header">
<th></th>
<th align="left">package</th>
<th align="left">loadedversion</th>
<th align="left">date</th>
<th align="left">source</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td align="left">assertthat</td>
<td align="left">0.2.0</td>
<td align="left">2017-04-11</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>2</td>
<td align="left">backports</td>
<td align="left">1.1.2</td>
<td align="left">2017-12-13</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>4</td>
<td align="left">bindr</td>
<td align="left">0.1</td>
<td align="left">2016-11-13</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>5</td>
<td align="left">bindrcpp</td>
<td align="left">0.2</td>
<td align="left">2017-06-17</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>6</td>
<td align="left">clisymbols</td>
<td align="left">1.2.0</td>
<td align="left">2017-05-21</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>7</td>
<td align="left">colorspace</td>
<td align="left">1.3-2</td>
<td align="left">2016-12-14</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>9</td>
<td align="left">cowplot</td>
<td align="left">0.9.2</td>
<td align="left">2017-12-17</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>11</td>
<td align="left">digest</td>
<td align="left">0.6.14</td>
<td align="left">2018-01-14</td>
<td align="left">cran (<span class="citation">@0.6.14</span>)</td>
</tr>
<tr class="odd">
<td>12</td>
<td align="left">dplyr</td>
<td align="left">0.7.4</td>
<td align="left">2017-09-28</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>13</td>
<td align="left">dropEstAnalysis</td>
<td align="left">0.6.0</td>
<td align="left">2018-01-27</td>
<td align="left">local (<a href="mailto:VPetukhov/dropEstAnalysis@NA">VPetukhov/dropEstAnalysis@NA</a>)</td>
</tr>
<tr class="odd">
<td>14</td>
<td align="left">dropestr</td>
<td align="left">0.6.0</td>
<td align="left">2018-01-24</td>
<td align="left">local (<span class="citation">@0.6.0</span>)</td>
</tr>
<tr class="even">
<td>15</td>
<td align="left">evaluate</td>
<td align="left">0.10.1</td>
<td align="left">2017-06-24</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>16</td>
<td align="left">ggplot2</td>
<td align="left">2.2.1</td>
<td align="left">2016-12-30</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>17</td>
<td align="left">ggrastr</td>
<td align="left">0.1.5</td>
<td align="left">2017-12-28</td>
<td align="left">Github (<a href="mailto:VPetukhov/ggrastr@cc56b45">VPetukhov/ggrastr@cc56b45</a>)</td>
</tr>
<tr class="odd">
<td>18</td>
<td align="left">git2r</td>
<td align="left">0.21.0</td>
<td align="left">2018-01-04</td>
<td align="left">cran (<span class="citation">@0.21.0</span>)</td>
</tr>
<tr class="even">
<td>19</td>
<td align="left">glue</td>
<td align="left">1.2.0</td>
<td align="left">2017-10-29</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>23</td>
<td align="left">gtable</td>
<td align="left">0.2.0</td>
<td align="left">2016-02-26</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>24</td>
<td align="left">highr</td>
<td align="left">0.6</td>
<td align="left">2016-05-09</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>25</td>
<td align="left">htmltools</td>
<td align="left">0.3.6</td>
<td align="left">2017-04-28</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>26</td>
<td align="left">knitr</td>
<td align="left">1.18</td>
<td align="left">2017-12-27</td>
<td align="left">cran (<span class="citation">@1.18</span>)</td>
</tr>
<tr class="odd">
<td>27</td>
<td align="left">labeling</td>
<td align="left">0.3</td>
<td align="left">2014-08-23</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>28</td>
<td align="left">lattice</td>
<td align="left">0.20-35</td>
<td align="left">2017-03-25</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>29</td>
<td align="left">lazyeval</td>
<td align="left">0.2.1</td>
<td align="left">2017-10-29</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>30</td>
<td align="left">magrittr</td>
<td align="left">1.5</td>
<td align="left">2014-11-22</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>31</td>
<td align="left">Matrix</td>
<td align="left">1.2-12</td>
<td align="left">2017-11-16</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>33</td>
<td align="left">munsell</td>
<td align="left">0.4.3</td>
<td align="left">2016-02-13</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>34</td>
<td align="left">pkgconfig</td>
<td align="left">2.0.1</td>
<td align="left">2017-03-21</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>35</td>
<td align="left">plyr</td>
<td align="left">1.8.4</td>
<td align="left">2016-06-08</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>36</td>
<td align="left">R6</td>
<td align="left">2.2.2</td>
<td align="left">2017-06-17</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>37</td>
<td align="left">Rcpp</td>
<td align="left">0.12.15</td>
<td align="left">2018-01-20</td>
<td align="left">cran (<span class="citation">@0.12.15</span>)</td>
</tr>
<tr class="odd">
<td>38</td>
<td align="left">rlang</td>
<td align="left">0.1.4</td>
<td align="left">2017-11-05</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>39</td>
<td align="left">rmarkdown</td>
<td align="left">1.8</td>
<td align="left">2017-11-17</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>40</td>
<td align="left">rprojroot</td>
<td align="left">1.3-2</td>
<td align="left">2018-01-03</td>
<td align="left">cran (<span class="citation">@1.3-2</span>)</td>
</tr>
<tr class="even">
<td>41</td>
<td align="left">scales</td>
<td align="left">0.5.0</td>
<td align="left">2017-08-24</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>42</td>
<td align="left">sessioninfo</td>
<td align="left">1.0.0</td>
<td align="left">2017-06-21</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>44</td>
<td align="left">stringi</td>
<td align="left">1.1.6</td>
<td align="left">2017-11-17</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>45</td>
<td align="left">stringr</td>
<td align="left">1.2.0</td>
<td align="left">2017-02-18</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="even">
<td>46</td>
<td align="left">tibble</td>
<td align="left">1.3.4</td>
<td align="left">2017-08-22</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
<tr class="odd">
<td>49</td>
<td align="left">withr</td>
<td align="left">2.1.1</td>
<td align="left">2017-12-19</td>
<td align="left">cran (<span class="citation">@2.1.1</span>)</td>
</tr>
<tr class="even">
<td>50</td>
<td align="left">yaml</td>
<td align="left">2.1.16</td>
<td align="left">2017-12-12</td>
<td align="left">CRAN (R 3.4.1)</td>
</tr>
</tbody>
</table>
</div>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
